### DISCUSSION TITLE
Regex order/options to allow beginning delimiter

### DISCUSSION THREAD

**Discussion by keeganjk** [beetbox/beets#4490] [Q&A] [answered]
Hello, my current goal for how files should look is number-track_name, that is, there is an initial number for the year or track, a hyphen as a delimiter, and then the title for the album or track. Is there any way I can perform a `replace` only on e.g. `$title` and `$album`, or is there a regular expression solution?

This is what my `replace` section currently looks like:
```
replace:
    '[\\/]': _
    '[\x00-\x1f]': _
    '[<>:\!\?\*\|_]': _
    '-+': _
    '^(\d+)\s+': '\1-'
    '\s+': _
    '[:,]': ''
    '[^a-z0-9_-]': ''
```

`'^(\d+)\s+': '\1-'` is supposed to preserve the initial number place a hyphen, but it looks like these hyphens are being replaced with underscores by `'-+': _`. What could I possibly do about this?

**Comment by sampsyo** [comment]
Hmm… that's an interesting problem; it's not exactly a use case that we've built the `replace` feature for. That is, its original purpose is to "legalize" filenames, in the sense that it matches illegal characters that the filesystem can't support well and replaces them with other stuff. Having one rule that _produces_ characters that another rule classifies as _illegal_ characters is not something we anticipated. The code for filename sanitation repeatedly applies the rules for exactly this kind of situation: where some rule might produce an invalid filename that needs to be further cleaned up by other rules.

I wonder if a custom plugin might be the best way to do this? Not entirely sure what a good route would be, to be honest, unless it's possible to make that `-+` rule more specific.

**Reply by keeganjk** [reply]
Interesting, that makes sense. I am trying to "legalize" this to strict regulations (maybe unnecessarily), that is, allowing only `[a-z0-9_.-]` where hyphen separates "items" (e.g. year or track number) and the dot is reserved for the file extension, so that leaves the underscore specifically for spaces within titles. Maybe doing this is more complicated than I thought, and/or unnecessary.

Though yes, if the regex I am currently trying is not possible, making a more detailed regex for `-+` (to fit the case where it is not preceeded by `^\d+`, which mostly works except for `+` within a reverse-lookback like `(?<!\d+)`) would be an ideal solution.

**Comment by wisp3rwind** [comment, marked as answer]
> Hello, my current goal for how files should look is number-track_name, that is, there is an initial number for the year or track, a hyphen as a delimiter, and then the title for the album or track. Is there any way I can perform a `replace` only on e.g. `$title` and `$album`, or is there a regular expression solution?
> 
> This is what my `replace` section currently looks like:
> 
> ```
> replace:
>     '[\\/]': _
>     '[\x00-\x1f]': _
>     '[<>:\!\?\*\|_]': _
>     '-+': _
>     '^(\d+)\s+': '\1-'
>     '\s+': _
>     '[:,]': ''
>     '[^a-z0-9_-]': ''
> ```
> 
> `'^(\d+)\s+': '\1-'` is supposed to preserve the initial number place a hyphen, but it looks like these hyphens are being replaced with underscores by `'-+': _`. What could I possibly do about this?

I'm having a hard time following what exactly you try to do here, since this is missing your configuration for context (in particular the `paths` section) and you don't really describe your actual goal.

That being said, this sounds like the [`inline` plugin](https://beets.readthedocs.io/en/stable/plugins/inline.html) should allow you to define sanitized fields, using any python code you like.

**Reply by keeganjk** [reply]
The rest of my configuration is quite simple; the only other field in use is `paths`. Here is the full scope:

```
directory: /media/hdd2/aud/
plugins: inline convert

paths: 
    default: %lower{%asciify{$albumartist/$year $album/$track $title}}
    comp: %lower{%asciify{compilations/$year $album%aunique{}/$track $title}}

replace:
    '[\\/]': _
    '[\x00-\x1f]': _
    '[<>:\!\?\*\|_]': _
    '-+': _
    '^(\d+)\s+': '\1-'
    '\s+': _
    '[:,]': ''
    '[^a-z0-9_-]': ''
```

I left spaces in `paths` as a "placeholder" that would separately be replaced with a hyphen.

My goal is to restrict filenames to simple ASCII characters: lowercase alphanumeric, the underscore for spaces (which, not being a delimiter, make sense for use within an item like title (for organizational reasons, and out of confidence that this will lead to no problems in compatibility or strange characters)), the dot being reserved for file extensions (not making much sense elsewhere), and the hyphen, a delimiter, being used to separate individual items (like track and title). Though maybe it does not make sense to do this, or it is useless (one should be able to simply tell that the first number indicates the track, though it seems to logically make sense to me to format this way).

`inline` seems promising for what I am trying to do; I will give this a shot.

**Reply by keeganjk** [reply]
I have included `inline` and the following seems to bring forth a solution for the problem I described:

```
directory: /media/hdd2/aud/
plugins: inline convert

# _f: _formatted
item_fields:
  album_f: album.replace('-', "_")
  title_f: title.replace('-', "_")
  albumartist_f: albumartist.replace('-', "_")

paths: 
  default: %lower{%asciify{$albumartist_f/$year-$album_f/$track-$title_f}}
  comp: %lower{%asciify{compilations/$year-$album_f%aunique{}/$track-$title_f}}

replace:
  '[\\/]': _
  '[\x00-\x1f]': _
  '[<>:\!\?\*\|_]': _
  '\s+': _
  '_+': _
  '[^a-z0-9_-]': ''
```
